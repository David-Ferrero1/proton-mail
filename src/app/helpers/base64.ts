import { encodeBase64, decodeBase64 } from 'pmcrypto';
import { arrayToBinaryString, binaryStringToArray } from 'proton-shared/lib/helpers/string';

export const arrayToBase64 = (data: Uint8Array): string => encodeBase64(arrayToBinaryString(data)) || '';

export const base64ToArray = (data: string): Uint8Array => binaryStringToArray(decodeBase64(data) || '');
