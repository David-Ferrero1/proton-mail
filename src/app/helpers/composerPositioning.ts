import { CSSProperties } from 'react';

import { WindowSize } from '../models/utils';

export const MAX_ACTIVE_COMPOSER_DESKTOP = 3;
export const MAX_ACTIVE_COMPOSER_MOBILE = 1;

const COMPOSER_WIDTH = 600;
const COMPOSER_HEIGHT = 520;
const COMPOSER_GUTTER = 20;
const COMPOSER_VERTICAL_GUTTER = 10;
const COMPOSER_ZINDEX = 300;
const COMPOSER_SWITCH_MODE = 20;
const HEADER_HEIGHT = 80;
const APP_BAR_WIDTH = 45;

const computeRightPosition = (index: number, count: number, windowWidth: number) => {
    const neededWidth = COMPOSER_WIDTH * count + COMPOSER_GUTTER * (count + 1);

    if (neededWidth < windowWidth) {
        return COMPOSER_WIDTH * index + COMPOSER_GUTTER * (index + 1);
    }

    const widthToDivide = windowWidth - COMPOSER_GUTTER * 2 - COMPOSER_WIDTH;
    const share = widthToDivide / (count - 1);
    return COMPOSER_GUTTER + share * index;
};

const computeHeight = (windowHeight: number) => {
    const maxHeight = windowHeight - COMPOSER_VERTICAL_GUTTER - HEADER_HEIGHT;
    return maxHeight > COMPOSER_HEIGHT ? COMPOSER_HEIGHT : maxHeight;
};

export const computeComposerStyle = (
    index: number,
    count: number,
    focus: boolean,
    minimized: boolean,
    maximized: boolean,
    isNarrow: boolean,
    windowSize: WindowSize
): CSSProperties => {
    if (isNarrow) {
        return {
            inset: 0,
            height: 'auto',
            width: 'auto'
        };
    }

    const style = {
        right: computeRightPosition(index, count, windowSize.width),
        zIndex: focus ? COMPOSER_ZINDEX + 1 : COMPOSER_ZINDEX,
        computeHeight: computeHeight(windowSize.height)
    };

    if (minimized) {
        return {
            ...style,
            height: 35
        };
    }

    if (maximized) {
        return {
            ...style,
            right: COMPOSER_GUTTER,
            width: windowSize.width - COMPOSER_GUTTER - APP_BAR_WIDTH,
            height: windowSize.height - COMPOSER_VERTICAL_GUTTER * 2
        };
    }

    return style;
};

export const shouldBeMaximized = (windowHeight: number) =>
    windowHeight - COMPOSER_VERTICAL_GUTTER - HEADER_HEIGHT < COMPOSER_HEIGHT - COMPOSER_SWITCH_MODE;
