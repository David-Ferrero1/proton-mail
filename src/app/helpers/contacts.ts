import { ContactEmailCache, ContactEmail } from 'proton-shared/lib/interfaces/contacts';
import { normalizeEmail } from './addresses';

export const findEmailInCache = (cache: ContactEmailCache, email: string): Partial<ContactEmail> =>
    [...cache.values()].find(({ Email }) => {
        return email === normalizeEmail(Email);
    }) || {};

export const getContactOfRecipient = (contacts: ContactEmail[] = [], address = '') =>
    (contacts || []).find(({ Email }) => normalizeEmail(Email) === normalizeEmail(address));

export const getContactsOfGroup = (contacts: ContactEmail[] = [], groupID = ''): ContactEmail[] =>
    contacts.filter((contact) => contact.LabelIDs?.includes(groupID));
