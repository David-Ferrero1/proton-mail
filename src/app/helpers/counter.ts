import { Counter } from '../models/counter';

export const replaceCounter = (counters: Counter[], counter: Counter) =>
    counters.map((existingCounter: any) => {
        if (existingCounter.LabelID === counter.LabelID) {
            return counter;
        }
        return existingCounter;
    });
