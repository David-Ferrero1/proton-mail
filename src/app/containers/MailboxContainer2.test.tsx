import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import MailboxContainer from './MailboxContainer';
import { Location, History } from 'history';
import { CacheProvider } from 'react-components';
import createCache from 'proton-shared/lib/helpers/cache';
import ApiContext from 'react-components/containers/api/apiContext';
import { wait } from 'proton-shared/lib/helpers/promise';
import { STATUS } from 'proton-shared/lib/models/cache';

// require.context
jest.mock('proton-shared/lib/i18n/dateFnLocales.js', () => ({}));
jest.mock('react-components/containers/payments/CardNumberInput.js', () => ({}));
jest.mock('react-components/containers/vpn/OpenVPNConfigurationSection/Country.js', () => ({}));

jest.mock('react-components/containers/eventManager/useEventManager.js', () => () => {
    return {
        subscribe: jest.fn()
    };
});

jest.mock('react-components/containers/notifications/useNotifications.js', () => () => {
    return {
        createNotification: jest.fn()
    };
});

jest.mock('react-components/containers/modals/useModals.js', () => () => {
    return {
        createModal: jest.fn()
    };
});

jest.mock('proton-shared/lib/helpers/setupPmcrypto.js', () => {
    return {
        initMain: jest.fn(),
        initWorker: jest.fn()
    };
});

jest.mock('./ConversationProvider');

const emptyProps = {
    labelID: 'labelID',
    mailSettings: {},
    elementID: 'elementID',
    location: {} as Location,
    history: {} as History,
    onCompose: jest.fn()
};

const api = jest.fn<Promise<any>, any>(async () => {
    // console.log('mocked api', ...args);
    return {};
});
const cache = createCache();

cache.set('User', { status: STATUS.RESOLVED, value: {} });

describe('MailboxContainer', () => {
    it('should match snapshot', async () => {
        const result = render(
            <ApiContext.Provider value={api}>
                <CacheProvider cache={cache}>
                    <MailboxContainer {...emptyProps} />
                </CacheProvider>
            </ApiContext.Provider>
        );
        await act(() => wait(0));

        expect(result).toMatchSnapshot();
    });

    it('should select all', async () => {
        cache.set('User', { status: STATUS.RESOLVED, value: {} });

        api.mockImplementation(async (args: any) => {
            if (args.url === 'conversations') {
                return {
                    Total: 2,
                    Conversations: [
                        { ID: '1', Labels: [{ ID: emptyProps.labelID }] },
                        { ID: '2', Labels: [{ ID: emptyProps.labelID }] }
                    ]
                };
            }

            return {};
        });

        const { container } = render(
            <ApiContext.Provider value={api}>
                <CacheProvider cache={cache}>
                    <MailboxContainer {...emptyProps} />
                </CacheProvider>
            </ApiContext.Provider>
        );
        await act(() => wait(0));

        const checkAll = container.querySelector('.toolbar .pm-checkbox') as Element;
        fireEvent.click(checkAll);

        const allChecks = container.querySelectorAll('input.item-checkbox') as NodeListOf<HTMLInputElement>;
        expect(allChecks.length > 0).toBe(true);

        const checked = [...allChecks].every((oneCheck) => oneCheck.checked);
        expect(checked).toBe(true);
    });
});
