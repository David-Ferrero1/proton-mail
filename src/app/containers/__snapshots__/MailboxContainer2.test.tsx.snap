// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailboxContainer should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="toolbar flex noprint flex-spacebetween"
      >
        <div
          class="flex"
        >
          <label
            class="inline-flex flex pl1"
          >
            <input
              class="pm-checkbox"
              type="checkbox"
            />
            <span
              class="pm-checkbox-fakecheck"
            >
              <svg
                class="icon-16p pm-checkbox-fakecheck-img"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-on"
                />
              </svg>
            </span>
          </label>
          <button
            aria-busy="false"
            aria-expanded="false"
            class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
            title="Open actions dropdown"
            type="button"
          >
            <span
              class="mauto"
            >
              <span
                class=""
              />
              <svg
                class="icon-12p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-caret"
                />
              </svg>
            </span>
          </button>
          <span
            class="toolbar-separator"
          />
          <button
            class="toolbar-button"
            title="Mark as read"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-read"
              />
            </svg>
          </button>
          <button
            class="toolbar-button"
            title="Mark as unread"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-unread"
              />
            </svg>
          </button>
          <span
            class="toolbar-separator"
          />
          <button
            class="toolbar-button"
            title="Move to trash"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-trash"
              />
            </svg>
          </button>
          <button
            class="toolbar-button"
            title="Move to inbox"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-inbox"
              />
            </svg>
          </button>
          <button
            class="toolbar-button"
            title="Move to archive"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-archive"
              />
            </svg>
          </button>
          <button
            class="toolbar-button"
            title="Move to spam"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-spam"
              />
            </svg>
          </button>
          <button
            class="toolbar-button"
            title="Empty folder"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-empty-folder"
              />
            </svg>
          </button>
          <span
            class="toolbar-separator"
          />
          <button
            aria-busy="false"
            aria-expanded="false"
            class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
            type="button"
          >
            <span
              class="mauto"
            >
              <span
                class="mr0-5"
              >
                <svg
                  class="icon-16p toolbar-icon"
                  focusable="false"
                  role="img"
                  viewBox="0 0 16 16"
                >
                  <use
                    xlink:href="#shape-folder"
                  />
                </svg>
              </span>
              <svg
                class="icon-12p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-caret"
                />
              </svg>
            </span>
          </button>
          <button
            aria-busy="false"
            aria-expanded="false"
            class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
            type="button"
          >
            <span
              class="mauto"
            >
              <span
                class="mr0-5"
              >
                <svg
                  class="icon-16p toolbar-icon"
                  focusable="false"
                  role="img"
                  viewBox="0 0 16 16"
                >
                  <use
                    xlink:href="#shape-label"
                  />
                </svg>
              </span>
              <svg
                class="icon-12p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-caret"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="flex"
        >
          <button
            aria-busy="false"
            aria-expanded="false"
            class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
            type="button"
          >
            <span
              class="mauto"
            >
              <span
                class="mr0-5"
              >
                <svg
                  class="icon-16p toolbar-icon"
                  focusable="false"
                  role="img"
                  viewBox="0 0 16 16"
                >
                  <use
                    xlink:href="#shape-bullet-points"
                  />
                </svg>
              </span>
              <svg
                class="icon-12p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-caret"
                />
              </svg>
            </span>
          </button>
          <button
            aria-busy="false"
            aria-expanded="false"
            class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
            type="button"
          >
            <span
              class="mauto"
            >
              <span
                class="mr0-5"
              >
                <svg
                  class="icon-16p toolbar-icon"
                  focusable="false"
                  role="img"
                  viewBox="0 0 16 16"
                >
                  <use
                    xlink:href="#shape-sort-new-old"
                  />
                </svg>
              </span>
              <svg
                class="icon-12p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-caret"
                />
              </svg>
            </span>
          </button>
          <button
            aria-busy="false"
            aria-expanded="false"
            class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
            title="Change layout"
            type="button"
          >
            <span
              class="mauto"
            >
              <span
                class="mr0-5"
              >
                <svg
                  class="icon-16p toolbar-icon"
                  focusable="false"
                  role="img"
                  viewBox="0 0 16 16"
                >
                  <use
                    xlink:href="#shape-layout-columns"
                  />
                </svg>
              </span>
              <svg
                class="icon-12p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-caret"
                />
              </svg>
            </span>
          </button>
          <span
            class="toolbar-separator"
          />
          <button
            class="toolbar-button"
            disabled=""
            title="Previous"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon rotateZ-90 mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </button>
          <button
            aria-busy="false"
            aria-expanded="false"
            class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
            disabled=""
            title="Change layout"
            type="button"
          >
            <span
              class="mauto"
            >
              <span
                class="mr0-5"
              >
                1
              </span>
              <svg
                class="icon-12p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-caret"
                />
              </svg>
            </span>
          </button>
          <button
            class="toolbar-button"
            disabled=""
            title="Next"
            type="button"
          >
            <svg
              class="icon-16p toolbar-icon rotateZ-270 mauto"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </button>
        </div>
      </nav>
      <div
        class="main-area--withToolbar flex-item-fluid flex reset4print"
      >
        <div
          class="items-column-list scroll-if-needed scroll-smooth-touch"
        >
          <div
            class="m2"
          >
            <figure
              class="flex-item-fluid aligncenter p3"
            >
              <img
                alt="No messages found"
                src="test-file-stub"
              />
              <figcaption
                class="mt2"
              >
                <h3
                  class="bold"
                >
                  No messages found
                </h3>
                <p
                  data-if="folder"
                >
                  You do not have any messages here
                </p>
              </figcaption>
            </figure>
          </div>
        </div>
        <section
          class="view-column-detail p2 flex-item-fluid scroll-if-needed"
        />
      </div>
    </div>
    <div
      aria-hidden="true"
      class="dropDown dropDown--bottom toolbar-dropdown"
      hidden=""
      id="dropdown-0"
      role="dialog"
      style="top: -1000px; left: -1000px;"
    >
      <div
        class="dropDown-content"
      >
        <ul
          class="unstyled mt0 mb0 ml1 mr1"
        >
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-show-all-emails"
                />
              </svg>
              Select all
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-unread"
                />
              </svg>
              All unread
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-read"
                />
              </svg>
              All read
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-star"
                />
              </svg>
              All unstarred
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-starfull"
                />
              </svg>
              All starred
            </button>
          </li>
        </ul>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="dropDown dropDown--bottom toolbar-dropdown"
      hidden=""
      id="dropdown-1"
      role="dialog"
      style="top: -1000px; left: -1000px;"
    />
    <div
      aria-hidden="true"
      class="dropDown dropDown--bottom toolbar-dropdown"
      hidden=""
      id="dropdown-2"
      role="dialog"
      style="top: -1000px; left: -1000px;"
    />
    <div
      aria-hidden="true"
      class="dropDown dropDown--bottom toolbar-dropdown"
      hidden=""
      id="dropdown-3"
      role="dialog"
      style="top: -1000px; left: -1000px;"
    >
      <div
        class="dropDown-content"
      >
        <ul
          class="unstyled mt0 mb0 ml1 mr1"
        >
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              disabled=""
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-bullet-points"
                />
              </svg>
              Show all
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-unread"
                />
              </svg>
              Show unread
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-read"
                />
              </svg>
              Show read
            </button>
          </li>
        </ul>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="dropDown dropDown--bottom toolbar-dropdown"
      hidden=""
      id="dropdown-4"
      role="dialog"
      style="top: -1000px; left: -1000px;"
    >
      <div
        class="dropDown-content"
      >
        <ul
          class="unstyled mt0 mb0 ml1 mr1"
        >
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-sort-small-large"
                />
              </svg>
              Size: small to large
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-sort-large-small"
                />
              </svg>
              Size: large to small
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              disabled=""
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-sort-new-old"
                />
              </svg>
              Date: new to old
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-sort-old-new"
                />
              </svg>
              Date: old to new
            </button>
          </li>
        </ul>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="dropDown dropDown--bottom toolbar-dropdown"
      hidden=""
      id="dropdown-5"
      role="dialog"
      style="top: -1000px; left: -1000px;"
    >
      <div
        class="dropDown-content"
      >
        <ul
          class="unstyled mt0 mb0 ml1 mr1"
        >
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              disabled=""
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-layout-columns"
                />
              </svg>
              Column layout
            </button>
          </li>
          <li
            class="dropDown-item"
          >
            <button
              class="w100 pt0-5 pb0-5 alignleft"
              type="button"
            >
              <svg
                class="icon-16p mr0-5"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-layout-rows"
                />
              </svg>
              Row layout
            </button>
          </li>
        </ul>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="dropDown dropDown--bottom dropDown--narrow toolbar-dropdown"
      hidden=""
      id="dropdown-6"
      role="dialog"
      style="top: -1000px; left: -1000px;"
    >
      <div
        class="dropDown-content"
      >
        <ul
          class="unstyled mt0 mb0 ml1 mr1"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <nav
      class="toolbar flex noprint flex-spacebetween"
    >
      <div
        class="flex"
      >
        <label
          class="inline-flex flex pl1"
        >
          <input
            class="pm-checkbox"
            type="checkbox"
          />
          <span
            class="pm-checkbox-fakecheck"
          >
            <svg
              class="icon-16p pm-checkbox-fakecheck-img"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-on"
              />
            </svg>
          </span>
        </label>
        <button
          aria-busy="false"
          aria-expanded="false"
          class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
          title="Open actions dropdown"
          type="button"
        >
          <span
            class="mauto"
          >
            <span
              class=""
            />
            <svg
              class="icon-12p toolbar-icon"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </span>
        </button>
        <span
          class="toolbar-separator"
        />
        <button
          class="toolbar-button"
          title="Mark as read"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-read"
            />
          </svg>
        </button>
        <button
          class="toolbar-button"
          title="Mark as unread"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-unread"
            />
          </svg>
        </button>
        <span
          class="toolbar-separator"
        />
        <button
          class="toolbar-button"
          title="Move to trash"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-trash"
            />
          </svg>
        </button>
        <button
          class="toolbar-button"
          title="Move to inbox"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-inbox"
            />
          </svg>
        </button>
        <button
          class="toolbar-button"
          title="Move to archive"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-archive"
            />
          </svg>
        </button>
        <button
          class="toolbar-button"
          title="Move to spam"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-spam"
            />
          </svg>
        </button>
        <button
          class="toolbar-button"
          title="Empty folder"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-empty-folder"
            />
          </svg>
        </button>
        <span
          class="toolbar-separator"
        />
        <button
          aria-busy="false"
          aria-expanded="false"
          class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
          type="button"
        >
          <span
            class="mauto"
          >
            <span
              class="mr0-5"
            >
              <svg
                class="icon-16p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-folder"
                />
              </svg>
            </span>
            <svg
              class="icon-12p toolbar-icon"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </span>
        </button>
        <button
          aria-busy="false"
          aria-expanded="false"
          class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
          type="button"
        >
          <span
            class="mauto"
          >
            <span
              class="mr0-5"
            >
              <svg
                class="icon-16p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-label"
                />
              </svg>
            </span>
            <svg
              class="icon-12p toolbar-icon"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="flex"
      >
        <button
          aria-busy="false"
          aria-expanded="false"
          class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
          type="button"
        >
          <span
            class="mauto"
          >
            <span
              class="mr0-5"
            >
              <svg
                class="icon-16p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-bullet-points"
                />
              </svg>
            </span>
            <svg
              class="icon-12p toolbar-icon"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </span>
        </button>
        <button
          aria-busy="false"
          aria-expanded="false"
          class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
          type="button"
        >
          <span
            class="mauto"
          >
            <span
              class="mr0-5"
            >
              <svg
                class="icon-16p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-sort-new-old"
                />
              </svg>
            </span>
            <svg
              class="icon-12p toolbar-icon"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </span>
        </button>
        <button
          aria-busy="false"
          aria-expanded="false"
          class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
          title="Change layout"
          type="button"
        >
          <span
            class="mauto"
          >
            <span
              class="mr0-5"
            >
              <svg
                class="icon-16p toolbar-icon"
                focusable="false"
                role="img"
                viewBox="0 0 16 16"
              >
                <use
                  xlink:href="#shape-layout-columns"
                />
              </svg>
            </span>
            <svg
              class="icon-12p toolbar-icon"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </span>
        </button>
        <span
          class="toolbar-separator"
        />
        <button
          class="toolbar-button"
          disabled=""
          title="Previous"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon rotateZ-90 mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-caret"
            />
          </svg>
        </button>
        <button
          aria-busy="false"
          aria-expanded="false"
          class="flex-item-noshrink flex-item-noshrink toolbar-button toolbar-button--dropdown"
          disabled=""
          title="Change layout"
          type="button"
        >
          <span
            class="mauto"
          >
            <span
              class="mr0-5"
            >
              1
            </span>
            <svg
              class="icon-12p toolbar-icon"
              focusable="false"
              role="img"
              viewBox="0 0 16 16"
            >
              <use
                xlink:href="#shape-caret"
              />
            </svg>
          </span>
        </button>
        <button
          class="toolbar-button"
          disabled=""
          title="Next"
          type="button"
        >
          <svg
            class="icon-16p toolbar-icon rotateZ-270 mauto"
            focusable="false"
            role="img"
            viewBox="0 0 16 16"
          >
            <use
              xlink:href="#shape-caret"
            />
          </svg>
        </button>
      </div>
    </nav>
    <div
      class="main-area--withToolbar flex-item-fluid flex reset4print"
    >
      <div
        class="items-column-list scroll-if-needed scroll-smooth-touch"
      >
        <div
          class="m2"
        >
          <figure
            class="flex-item-fluid aligncenter p3"
          >
            <img
              alt="No messages found"
              src="test-file-stub"
            />
            <figcaption
              class="mt2"
            >
              <h3
                class="bold"
              >
                No messages found
              </h3>
              <p
                data-if="folder"
              >
                You do not have any messages here
              </p>
            </figcaption>
          </figure>
        </div>
      </div>
      <section
        class="view-column-detail p2 flex-item-fluid scroll-if-needed"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
